File: snow_2.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(year_n)
  DATA_INTEGER(size_n)
  DATA_MATRIX(imm_n_at_size_obs)
  DATA_MATRIX(mat_n_at_size_obs)
  DATA_MATRIX(prob_term_molt)
  DATA_MATRIX(size_trans)
  DATA_SCALAR(log_mu_m)
  DATA_SCALAR(sigma_numbers_imm)
  DATA_SCALAR(sigma_numbers_mat)
  DATA_SCALAR(sigma_m)

  PARAMETER_VECTOR(log_n_imm);
  PARAMETER_VECTOR(log_n_mat);
  PARAMETER_MATRIX(nat_m);
  PARAMETER_VECTOR(log_recruits);
  PARAMETER(logit_prop_rec);
  // try this with natural mortality for immature and mature animals split...molting could be an issue

  matrix<Type> imm_n_at_size_pred(year_n+1,size_n);
  matrix<Type> mat_n_at_size_pred(year_n+1,size_n);
  vector<Type> temp_imm(size_n);
  vector<Type> temp_mat(size_n);
  vector<Type> trans_imm(size_n);
 
  vector<Type> imm_numbers_obs(year_n);
  vector<Type> mat_numbers_obs(year_n);
  vector<Type> imm_numbers_pred(year_n);
  vector<Type> mat_numbers_pred(year_n);
 
  Type prop_rec;
  Type imm_num_like;
  Type mat_num_like;
  Type imm_like;
  Type mat_like;
  Type nat_m_like;
  Type obj_fun;

  // End of specifications section
  // =============================

  prop_rec = 1 / (1 + exp(logit_prop_rec));

  // initial year numbers at size
  for (int size=0;size<size_n;size++) 
  {
	imm_n_at_size_pred(0,size) = exp(log_n_imm(size));
	mat_n_at_size_pred(0,size) = exp(log_n_mat(size));
  }
  
  // Project each numbers at size matrix
   for (int year=0;year<year_n;year++) 
   {
	  // molting, growth, and mortality (not sure why this didn't work...)
      // imm_n_at_size_pred.row(year+1) = imm_n_at_size_pred.row(year) * (1-prob_term_molt.row(year)) * exp(-1*nat_m.row(year)) * size_trans;
      // mat_n_at_size_pred.row(year+1) = (imm_n_at_size_pred.row(year) * (prob_term_molt.row(year)) * exp(-1*nat_m.row(year)) * size_trans) + mat_n_at_size_pred.row(year) * exp(-1*nat_m.row(year));

     // natural mortality
      for (int size=0;size<size_n;size++) 
	   {
		temp_imm(size) = imm_n_at_size_pred(year,size) * exp(-1*exp(nat_m(year,size)));
		temp_mat(size) = mat_n_at_size_pred(year,size) * exp(-1*exp(nat_m(year,size)));
	   }	
	  
	  // growth...why doesn't the first one worK?
	   trans_imm = size_trans * temp_imm;
	  
	  // maturity
	   for (int size=0;size<size_n;size++) 
	   {
	  	 imm_n_at_size_pred(year+1,size) = trans_imm(size) * (1-prob_term_molt(year,size));	
	  	 mat_n_at_size_pred(year+1,size) = trans_imm(size) * prob_term_molt(year,size) + temp_mat(size);
       }
	  
	  // add recruitment
	   imm_n_at_size_pred(year+1,0) = exp(log_recruits(year))*prop_rec;
	   imm_n_at_size_pred(year+1,1) = exp(log_recruits(year))*(1-prop_rec);
   }
  
  
  // Likelihood components
  // immature numbers at size data
  
    // make total numbers by maturity state from obs and preds
  imm_numbers_obs = 0;
  mat_numbers_obs = 0;
  imm_numbers_pred = 0;
  mat_numbers_pred = 0;
  
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
   {
    imm_numbers_pred(year) += imm_n_at_size_pred(year,size);
    mat_numbers_pred(year) += mat_n_at_size_pred(year,size);
	imm_numbers_obs(year) += imm_n_at_size_obs(year,size);
    mat_numbers_obs(year) += mat_n_at_size_obs(year,size); 
   }  

  // total numbers likelihoods
  imm_num_like = 0;
  for (int year=0;year<year_n;year++)
   if (imm_n_at_size_obs(year,1) >0)
    imm_num_like += square( log(imm_numbers_pred(year)) - log(imm_numbers_obs(year))) / (2.0 * square(sigma_numbers_imm));
 
  mat_num_like = 0;
  for (int year=0;year<year_n;year++)
   if (mat_n_at_size_obs(year,1) >0)
    mat_num_like += square( log(mat_numbers_pred(year)) - log(mat_numbers_obs(year))) / (2.0 * square(sigma_numbers_mat));

 
  // immature numbers size composition
  imm_like = 0;
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
    if (imm_n_at_size_obs(year,size) >0)
     imm_like += (imm_n_at_size_obs(year,size)/imm_numbers_obs(year)) * log( (imm_n_at_size_pred(year,size)/imm_numbers_pred(year)) / (imm_n_at_size_obs(year,size)/imm_numbers_obs(year)));
  imm_like = -1*imm_like;
  
  // mature numbers size composition
  mat_like = 0;
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
    if (mat_n_at_size_obs(year,size) >0)
     mat_like += (mat_n_at_size_obs(year,size)/mat_numbers_obs(year)) * log( (mat_n_at_size_pred(year,size)/mat_numbers_pred(year)) / (mat_n_at_size_obs(year,size)/mat_numbers_obs(year)));
  mat_like = -1*mat_like;

  // natural mortality likelihood
  nat_m_like =0;
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
   nat_m_like += pow((nat_m(year,size)-log_mu_m)/ (sqrt(2)*sqrt(exp(sigma_m))),2.0);

  obj_fun = imm_num_like + mat_num_like + imm_like + mat_like + nat_m_like;
  
  REPORT(imm_n_at_size_pred);
  REPORT(mat_n_at_size_pred);
  REPORT(imm_numbers_pred);
  REPORT(mat_numbers_pred);
  REPORT(nat_m);
  REPORT(log_recruits);
  REPORT(imm_num_like);
  REPORT(mat_num_like);
  REPORT(imm_like);
  REPORT(mat_like);
  REPORT(nat_m_like);
  REPORT(obj_fun);


  return(obj_fun);

}

File: snow_2.R
#==========================================================================================
# try to estimate sigma_m
# try to estimate log_m_mu
# input sigma for total numbers
require(TMB)
library(reshape2)
library(ggplot2)
library(ggridges)
setwd("/tmb")
compile("snow_2.cpp") 
dyn.load(dynlib("snow_2"))


#gdbsource("models/snow.dll")

################################################################################
# INPUT DATA
############
year_n <- scan("dat_out.DAT",skip=2,n=1,quiet=T)
years <- scan("dat_out.DAT",skip=4,n=year_n,quiet=T)
size_n <- scan("dat_out.DAT",skip=6,n=1,quiet=T)
sizes <- scan("dat_out.DAT",skip=8,n=size_n,quiet=T)

imm_n_at_size_obs <- matrix(scan("dat_out.DAT",skip=10,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)
mat_n_at_size_obs <- matrix(scan("dat_out.DAT",skip=44,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)

prob_term_molt <- matrix(scan("dat_out.DAT",skip=78,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)
size_trans <- (matrix(scan("dat_out.DAT",skip=112,n=size_n*size_n,quiet=T),ncol=size_n,byrow=T))


################################################################################
# INPUT INITIAL PARAMETER VALUES
################################
log_n_imm <- scan("dat_out.PIN",skip=2,n=size_n,quiet=T)
log_n_mat <- scan("dat_out.PIN",skip=4,n=size_n,quiet=T)
nat_m <- matrix(scan("dat_out.PIN",skip=6,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)
log_recruits <- scan("dat_out.PIN",skip=40,n=year_n,quiet=T)
log_recruits <- log(imm_n_at_size_obs[,1])
log_recruits[log_recruits<0]<-0
log_mu_m <- -1.2
sigma_m <- 1.5
sigma_numbers_imm <- 0.2
sigma_numbers_mat <- 0.2
logit_prop_rec <- 0
################################################################################
# Data vector
data <- list(year_n=year_n,size_n=size_n,imm_n_at_size_obs=imm_n_at_size_obs,
             mat_n_at_size_obs=mat_n_at_size_obs,prob_term_molt=prob_term_molt,
             size_trans=size_trans,log_mu_m=log_mu_m,
             sigma_numbers_imm=sigma_numbers_imm,sigma_numbers_mat=sigma_numbers_mat,
             sigma_m=sigma_m)

parameters <- list(log_n_imm=log(imm_n_at_size_obs[1,]),log_n_mat=log(mat_n_at_size_obs[1,]),
                   nat_m=nat_m,log_recruits=log_recruits,
                   logit_prop_rec=logit_prop_rec)

# map<-list(nat_m=matrix(factor(NA),ncol=ncol(nat_m),nrow=nrow(nat_m)))
# map<-list(sigma_m=factor(NA),log_mu_m=factor(NA))
# map<-list(sigma_m=factor(NA),log_mu_m=factor(NA),
#           log_n_imm=rep(factor(NA),length(log_n_imm)),
#           log_n_mat=rep(factor(NA),length(log_n_imm)))
# map<-list(sigma_m=factor(NA))
#map<-list(log_n_imm=rep(factor(NA),length(log_n_imm)),
#          log_n_mat=rep(factor(NA),length(log_n_imm)),
#          recruits=rep(factor(NA),length(recruits)))
#print(data)
#print(parameters)

################################################################################
model <- MakeADFun(data, parameters, DLL="snow_2",silent=T,random='nat_m')
model <- MakeADFun(data, parameters, DLL="snow_2",silent=T)

# test code - for checking for minimization
xx <- model$fn(model$env$last.par)
#print(model$report())
#cat(model$report()$obj_fun,model$report()$Like1,model$report()$Like2,model$report()$Like3,"\n")

# Actual minimzation (with some "Bonus" parameters from nlminb)
fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=100000,iter.max=10000))
#fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=1,iter.max=1))
best <- model$env$last.par.best
rep <- sdreport(model)
print(best)
print(rep)
print(model$report()$imm_like)
print(model$report()$mat_like)
print(model$report()$nat_m_like)
cat(model$report()$obj_fun,model$report()$Like1,model$report()$Like2,model$report()$Like3,"\n")
rep <- sdreport(model)


#=========================================================
# model fits and estimated parameters
#+========================================================

#=============================================
# plot immature and mature numbers by year
#=============================================
tot_mat_obs<-apply(mat_n_at_size_obs,1,sum)
div_n<-1000000000
par(mfrow=c(2,1),mar=c(.1,.1,.3,.1),oma=c(4,4,1,4))
plot_mmb<-tot_mat_obs/div_n
plot_mmb[plot_mmb==0]<-NA

plot(plot_mmb~years,
     pch=20,ylab="Mature male abundance",
     xlab="Year",las=1,ylim=c(0,2),xaxt='n')

for(j in 1:length(years))
{
  segments(x0=years[j],x1=years[j],
           y0=plot_mmb[j] /  exp(1.96*sqrt(log(1+0.2^2))),
           y1=plot_mmb[j] *  exp(1.96*sqrt(log(1+0.2^2))))
}

lines(model$report()$mat_numbers_pred/div_n ~ years,lwd=2)
legend('topleft',bty='n',"Mature males")



tot_imm_obs<-apply(imm_n_at_size_obs,1,sum)
plot_mmb<-tot_imm_obs/div_n
plot_mmb[plot_mmb==0]<-NA

plot(plot_mmb~years,
     pch=20,ylab="Mature male abundance",
     xlab="Year",las=1,ylim=c(0,5))

for(j in 1:length(years))
{
  segments(x0=years[j],x1=years[j],
           y0=plot_mmb[j] /  exp(1.96*sqrt(log(1+0.2^2))),
           y1=plot_mmb[j] *  exp(1.96*sqrt(log(1+0.2^2))))
}
lines(model$report()$imm_numbers_pred/div_n ~ years,lwd=2)
legend('topleft',bty='n',"Immature males")
mtext(side=2,line=2.5,"Abundance (billions)",outer=T)

#==size comps



par(mfrow=c(6,6),mar=c(.1,.1,.1,.1),oma=c(4,4,1,1))
for(x in 1:nrow(imm_n_at_size_obs))
{
  plot(imm_n_at_size_obs[x,],type='l',ylim=c(0,max(imm_n_at_size_obs[x,],model$report()$imm_n_at_size_pred[x,])),
       yaxt='n',xaxt='n')
  lines(model$report()$imm_n_at_size_pred[x,],lty=2,col=2)
  legend('topright',bty='n',legend=years[x])
}

par(mfrow=c(6,6),mar=c(.1,.1,.1,.1),oma=c(4,4,1,1))
for(x in 1:nrow(imm_n_at_size_obs))
{
  plot(mat_n_at_size_obs[x,],type='l',ylim=c(0,max(mat_n_at_size_obs[x,],model$report()$mat_n_at_size_pred[x,])),
       yaxt='n',xaxt='n')
  lines(model$report()$imm_n_at_size_pred[x,],lty=2,col=2)
  legend('topright',bty='n',legend=years[x])
}

plot_m<-exp(model$report()$nat_m)
colnames(plot_m)<-sizes
rownames(plot_m)<-years
in_dat<-melt(plot_m)
colnames(in_dat)<-c("Year","Size","nat_m")

size_dat<- ggplot(dat=in_dat) 
size_dat <- size_dat + geom_density_ridges(aes(x=Size, y=Year, height = nat_m,
                                               group = Year, 
                                               fill=stat(y),alpha=.9999),stat = "identity",scale=1) +
  # scale_fill_viridis_c()+
  theme_bw() +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 0),
        axis.title.y=element_blank(),
        axis.title.x=element_blank()) 
print(size_dat)
